File: behavioral/typescript/interpreter_structured/arithmetic_expressions.ts
import { Expression } from './interfaces';

export class NumberExpression implements Expression {
  private value: number;
  
  constructor(value: number) {
    this.value = value;
  }
  
  interpret(): number {
    return this.value;
  }
}

export class AddExpression implements Expression {
  private left: Expression;
  private right: Expression;
  
  constructor(left: Expression, right: Expression) {
    this.left = left;
    this.right = right;
  }
  
  interpret(): number {
    return this.left.interpret() + this.right.interpret();
  }
}

export class SubtractExpression implements Expression {
  private left: Expression;
  private right: Expression;
  
  constructor(left: Expression, right: Expression) {
    this.left = left;
    this.right = right;
  }
  
  interpret(): number {
    return this.left.interpret() - this.right.interpret();
  }
}

export class MultiplyExpression implements Expression {
  private left: Expression;
  private right: Expression;
  
  constructor(left: Expression, right: Expression) {
    this.left = left;
    this.right = right;
  }
  
  interpret(): number {
    return this.left.interpret() * this.right.interpret();
  }
}

export class ExpressionParser {
  private expression: string;
  private pos: number = 0;
  
  constructor(expr: string) {
    this.expression = expr;
  }
  
  private peek(): string {
    return this.pos < this.expression.length ? this.expression[this.pos] : '\0';
  }
  
  private next(): string {
    return this.expression[this.pos++];
  }
  
  private skipWhitespace(): void {
    while (this.pos < this.expression.length && /\s/.test(this.expression[this.pos])) {
      this.pos++;
    }
  }
  
  private parseExpression(): Expression {
    let left = this.parseTerm();
    
    while (this.peek() === '+' || this.peek() === '-') {
      const op = this.next();
      const right = this.parseTerm();
      
      if (op === '+') {
        left = new AddExpression(left, right);
      } else {
        left = new SubtractExpression(left, right);
      }
    }
    
    return left;
  }
  
  private parseTerm(): Expression {
    let left = this.parseFactor();
    
    while (this.peek() === '*' || this.peek() === '/') {
      const op = this.next();
      const right = this.parseFactor();
      
      if (op === '*') {
        left = new MultiplyExpression(left, right);
      }
    }
    
    return left;
  }
  
  private parseFactor(): Expression {
    this.skipWhitespace();
    
    if (this.peek() === '(') {
      this.next();
      const expr = this.parseExpression();
      this.next();
      return expr;
    }
    
    this.skipWhitespace();
    let value = 0;
    while (this.pos < this.expression.length && /\d/.test(this.expression[this.pos])) {
      value = value * 10 + parseInt(this.expression[this.pos]);
      this.pos++;
    }
    
    return new NumberExpression(value);
  }
  
  parse(): Expression {
    return this.parseExpression();
  }
}

File: behavioral/dart/interpreter_structured/arithmetic_expressions.dart
import 'interfaces.dart';

class NumberExpression extends Expression {
  final int _value;
  
  NumberExpression(this._value);
  
  @override
  int interpret() {
    return _value;
  }
}

class AddExpression extends Expression {
  final Expression _left;
  final Expression _right;
  
  AddExpression(this._left, this._right);
  
  @override
  int interpret() {
    return _left.interpret() + _right.interpret();
  }
}

class SubtractExpression extends Expression {
  final Expression _left;
  final Expression _right;
  
  SubtractExpression(this._left, this._right);
  
  @override
  int interpret() {
    return _left.interpret() - _right.interpret();
  }
}

class MultiplyExpression extends Expression {
  final Expression _left;
  final Expression _right;
  
  MultiplyExpression(this._left, this._right);
  
  @override
  int interpret() {
    return _left.interpret() * _right.interpret();
  }
}

class ExpressionParser {
  final String _expression;
  int _pos = 0;
  
  ExpressionParser(this._expression);
  
  String _peek() {
    return _pos < _expression.length ? _expression[_pos] : '\0';
  }
  
  String _next() {
    return _expression[_pos++];
  }
  
  void _skipWhitespace() {
    while (_pos < _expression.length && RegExp(r'\s').hasMatch(_expression[_pos])) {
      _pos++;
    }
  }
  
  Expression _parseExpression() {
    var left = _parseTerm();
    
    while (_peek() == '+' || _peek() == '-') {
      final op = _next();
      final right = _parseTerm();
      
      if (op == '+') {
        left = AddExpression(left, right);
      } else {
        left = SubtractExpression(left, right);
      }
    }
    
    return left;
  }
  
  Expression _parseTerm() {
    var left = _parseFactor();
    
    while (_peek() == '*' || _peek() == '/') {
      final op = _next();
      final right = _parseFactor();
      
      if (op == '*') {
        left = MultiplyExpression(left, right);
      }
    }
    
    return left;
  }
  
  Expression _parseFactor() {
    _skipWhitespace();
    
    if (_peek() == '(') {
      _next();
      final expr = _parseExpression();
      _next();
      return expr;
    }
    
    _skipWhitespace();
    var value = 0;
    while (_pos < _expression.length && RegExp(r'\d').hasMatch(_expression[_pos])) {
      value = value * 10 + int.parse(_expression[_pos]);
      _pos++;
    }
    
    return NumberExpression(value);
  }
  
  Expression parse() {
    return _parseExpression();
  }
}
